pkg/controller/cloudinit/controller.go

```diff
@@ -0,0 +1,403 @@
+	}
+
+	if cloudInitCopy.Spec.Paused {
+		return c.updateStatus(node, cloudInitCopy)
+	}
+
+	if cloudinit.MatchesNode(node, cloudInitCopy) {
+		changed, err := cloudinit.RequireLocal(cloudInitCopy)
+		_, _ = c.updateStatus(node, cloudInitCopy)
+		if err != nil {
+			return cloudInitCopy, err
+		}
+
+		if !changed {
+			return cloudInitCopy, nil
+		}
+
+		err = c.emitOverwriteEvent(cloudInitCopy)
+		if err != nil {
+			logrus.WithError(err).
+				WithField("cloudinit_name", cloudInitCopy.Name).
+				Warn("Failed to emit overwrite event for CloudInit")
+		}
+
+		return cloudInitCopy, err
+	}
+
+	err = os.Remove(filepath.Join(cloudinit.Directory, cloudInitCopy.Spec.Filename))
+	if os.IsNotExist(err) {
+		return c.updateStatus(node, cloudInitCopy)
+	}
+
+	err = c.emitRemoveEvent(cloudInitCopy)
+	if err != nil {
+		logrus.WithError(err).
+			WithField("cloudinit_name", cloudInitCopy.Name).
+			Warn("Failed to emit removal event for CloudInit")
+	}
+
+	return c.updateStatus(node, cloudInitCopy)
+}
+
+func (c *controller) OnCloudInitRemove(_ string, cloudInitObj *cloudinitv1.CloudInit) (*cloudinitv1.CloudInit, error) {
+	err := os.Remove(filepath.Join(cloudinit.Directory, cloudInitObj.Spec.Filename))
+	if os.IsNotExist(err) {
+		return cloudInitObj, nil
+	}
+
+	if err != nil {
+		return cloudInitObj, err
+	}
+
+	return cloudInitObj, nil
+}
+
+type byCondType []metav1.Condition
+
+func (n byCondType) Len() int           { return len(n) }
+func (n byCondType) Swap(i, j int)      { n[i], n[j] = n[j], n[i] }
+func (n byCondType) Less(i, j int) bool { return n[i].Type < n[j].Type }
+
+const (
+	CloudInitReasonPresent          = "CloudInitPresentOnDisk"
+	CloudInitReasonAbsent           = "CloudInitAbsentFromDisk"
+	CloudInitReasonError            = "CloudInitError"
+	CloudInitReasonApplicable       = "CloudInitApplicable"
+	CloudInitReasonNotApplicable    = "CloudInitNotApplicable"
+	CloudInitReasonChecksumMismatch = "CloudInitChecksumMismatch"
+	CloudInitReasonChecksumMatch    = "CloudInitChecksumMatch"
+)
+
+func (c *controller) updateStatus(node *corev1.Node, cloudInitObj *cloudinitv1.CloudInit) (*cloudinitv1.CloudInit, error) {
+	var st unix.Stat_t
+	_ = unix.Stat(filepath.Join(cloudinit.Directory, cloudInitObj.Spec.Filename), &st)
+	createdAt := time.Unix(st.Ctim.Sec, st.Ctim.Nsec)
+	modifiedAt := time.Unix(st.Mtim.Sec, st.Mtim.Nsec)
+
+	conds := []metav1.Condition{
+		newApplicableCondition(node, cloudInitObj),
+		newOutOfSyncCondition(cloudInitObj),
+		newPresentCondition(cloudInitObj),
+	}
+	sort.Sort(byCondType(conds))
+
+	rollout := cloudInitObj.Status.Rollouts[c.nodeName]
+
+	oldConds := make(map[string]metav1.Condition)
+	for _, cond := range rollout.Conditions {
+		oldConds[cond.Type] = cond
+	}
+
+	now := time.Now()
+	for i := range conds {
+		cond := &conds[i]
+		prev := oldConds[cond.Type]
+
+		switch cond.Type {
+		case string(cloudinitv1.CloudInitFilePresent):
+			if createdAt.After(cond.LastTransitionTime.Time) {
```

Comment:
Why do we use `createdAt` instead of the current time?
I think that should not have been much different because you just init createdAt when call into this func

Reply:
oops... missing this line
```
_ = unix.Stat(filepath.Join(cloudinit.Directory, cloudInitObj.Spec.Filename), &st)
```

You are right, that make sense to me!